### PR TITLE
[PROJ] update to 8.2

### PR DESCRIPTION
**Pull request by visr**
Let's see what CI says. Locally I only got an error with `x86_64-linux-musl`:

```
[21:29:23] cd /workspace/srcdir/proj-8.2.0/build/src && /usr/bin/cmake -E cmake_link_script CMakeFiles/binproj.dir/link.txt --verbose=true
[21:29:23] /opt/bin/x86_64-linux-musl-libgfortran3-cxx03/x86_64-linux-musl-g++ --sysroot=/opt/x86_64-linux-musl/x86_64-linux-musl/sys-root/  -O3 -DNDEBUG   CMakeFiles/binproj.dir/apps/proj.cpp.o CMakeFiles/binproj.dir/apps/emess.cpp.o CMakeFiles/binproj.dir/apps/utils.cpp.o  -o ../bin/proj   -L/workspace/srcdir/proj-8.2.0/build/lib  -Wl,-rpath,/workspace/srcdir/proj-8.2.0/build/lib: ../lib/libproj.so.22.2.0 
[21:29:23] ../lib/libproj.so.22.2.0: undefined reference to `curl_easy_perform@CURL_4'
[21:29:23] ../lib/libproj.so.22.2.0: undefined reference to `curl_easy_init@CURL_4'
[21:29:23] ../lib/libproj.so.22.2.0: undefined reference to `curl_easy_getinfo@CURL_4'
[21:29:23] ../lib/libproj.so.22.2.0: undefined reference to `curl_easy_setopt@CURL_4'
[21:29:23] ../lib/libproj.so.22.2.0: undefined reference to `curl_easy_cleanup@CURL_4'
[21:29:23] collect2: error: ld returned 1 exit status
[21:29:23] make[2]: *** [src/CMakeFiles/binproj.dir/build.make:135: bin/proj] Error 1
[21:29:23] make[1]: *** [CMakeFiles/Makefile2:404: src/CMakeFiles/binproj.dir/all] Error 2
```

Though I'm not sure what to do about it, since we pass CURL_INCLUDE_DIR and CURL_LIBRARY.

I think it's fine to require julia 1.6 for new PROJ builds.

### CODE REVIEW

File: P/PROJ/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+products = [
+    LibraryProduct(["libproj", "libproj_$(upstream_version.major)_$(upstream_version.minor)"], :libproj),
+
+    ExecutableProduct("proj", :proj),
+    ExecutableProduct("gie", :gie),
+    ExecutableProduct("projinfo", :projinfo),
+    ExecutableProduct("projsync", :projsync),
+    ExecutableProduct("cs2cs", :cs2cs),
+    ExecutableProduct("geod", :geod),
+    ExecutableProduct("cct", :cct)
```
Comment (giordano) [inline P/PROJ/build_tarballs.jl:67]:
```suggestion
    ExecutableProduct("cct", :cct),
```

**Comment by giordano** [comment]
For the record, I just built the current Julia v1.6 recipe for `x86_64-linux-musl-cxx11` and it worked fine, so at least we didn't break anything in the meantime.

**Comment by giordano** [comment]
Of course, manually adding `-lcurl` to the compiler invocation would solve the issue.  `libproj.so` depends on libcurl:
```console
sandbox:${WORKSPACE}/srcdir/proj-8.2.0/build # readelf -d lib/libproj.so.22.2.0 | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libsqlite3.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libtiff.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libcurl.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.musl-x86_64.so.1]
```
The question is: what's happening of different on this platform?

**Comment by giordano** [comment]
Interestingly enough, the compiler invocation on `x86_64-linux-gnu-cxx11` looks identical:
```
[100%] Linking CXX executable ../bin/projsync
cd /workspace/srcdir/proj-8.2.0/build/src && /usr/bin/cmake -E cmake_link_script CMakeFiles/bin_projsync.dir/link.txt --verbose=true
/opt/bin/x86_64-linux-gnu-libgfortran3-cxx11/x86_64-linux-gnu-g++ --sysroot=/opt/x86_64-linux-gnu/x86_64-linux-gnu/sys-root/  -O3 -DNDEBUG   CMakeFiles/bin_projsync.dir/apps/projsync.cpp.o  -o ../bin/projsync   -L/workspace/srcdir/proj-8.2.0/build/lib  -Wl,-rpath,/workspace/srcdir/proj-8.2.0/build/lib: ../lib/libproj.so.22.2.0 
````
and also on this platform `libproj.so` depends on libcurl:
```console
% readelf -d x86_64-linux-gnu-cxx11/lib/libproj.so | grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libsqlite3.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [libtiff.so.5]
 0x0000000000000001 (NEEDED)             Shared library: [libcurl.so.4]
 0x0000000000000001 (NEEDED)             Shared library: [libstdc++.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
```
Sigh.

**Comment by visr** [comment]
Thanks for checking it out @giordano. I'd be fine with adding `-lcurl` for `x86_64-linux-musl`. How should that be done? Something like this?
https://github.com/JuliaPackaging/Yggdrasil/blob/dd1623bc3966f4503620dc9bf4a575f4f96cdfd8/S/samtools/build_tarballs.jl#L19-L22
If you suspect this might be an issue with the PROJ build I could ask on the PROJ mailing list. I see two musl issues on https://github.com/OSGeo/PROJ but there is no CI for it as far as I can tell.

**Comment by giordano** [comment]
That might work around the issue, yes, although I'm really not sure why the error is happening in the first place, and why only on one platform

**Comment by giordano** [comment]
I'm honestly quite lost.  When linking `bin/projsync` the linker on `x86_64-linux-musl`, and only on this platform, is trying to resolve undefined symbols in `lib/libproj.so` (which had been linked successfully earlier).  This doesn't seem to happen anywhere else, the linker doesn't use libcurl anywhere when linking `bin/projsync`, this doesn't make much sense to me.

**Comment by giordano** [comment]
For the record, log of building PROJ 7.2 for the same platform:
```
[20:17:54] [100%] Linking CXX executable ../bin/projsync
[20:17:54] cd /workspace/srcdir/proj-7.2.1/build/src && /usr/bin/cmake -E cmake_link_script CMakeFiles/bin_projsync.dir/link.txt --verbose=true
[20:17:54] /opt/bin/x86_64-linux-musl-libgfortran3-cxx11/x86_64-linux-musl-g++ --sysroot=/opt/x86_64-linux-musl/x86_64-linux-musl/sys-root/  -O3 -DNDEBUG  -Wl,-t CMakeFiles/bin_projsync.dir/apps/projsync.cpp.o  -o ../bin/projsync   -L/workspace/srcdir/proj-7.2.1/build/lib  -Wl,-rpath,/workspace/srcdir/proj-7.2.1/build/lib:/workspace/destdir/lib: ../lib/libproj.so.21.1.2 -lm -ldl /workspace/destdir/lib/libsqlite3.so /workspace/destdir/lib/libtiff.so /workspace/destdir/lib/libcurl.so 
[20:17:54] ccache /opt/x86_64-linux-musl/bin/x86_64-linux-musl-g++ -D_GLIBCXX_USE_CXX11_ABI=1 -frandom-seed=0x4874e8a1 -L/opt/x86_64-linux-musl/x86_64-linux-musl/lib64 -Wl,-rpath-link,/opt/x86_64-linux-musl/x86_64-linux-musl/lib64 -march=x86-64 -mtune=generic --sysroot=/opt/x86_64-linux-musl/x86_64-linux-musl/sys-root/ -O3 -DNDEBUG -Wl,-t CMakeFiles/bin_projsync.dir/apps/projsync.cpp.o -o ../bin/projsync -L/workspace/srcdir/proj-7.2.1/build/lib -Wl,-rpath,/workspace/srcdir/proj-7.2.1/build/lib:/workspace/destdir/lib: ../lib/libproj.so.21.1.2 -lm -ldl /workspace/destdir/lib/libsqlite3.so /workspace/destdir/lib/libtiff.so /workspace/destdir/lib/libcurl.so
[20:17:54] /opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/5.2.0/../../../../x86_64-linux-musl/bin/ld: mode elf_x86_64
[20:17:54] /opt/x86_64-linux-musl/x86_64-linux-musl/sys-root/usr/lib/crt1.o
[20:17:54] /opt/x86_64-linux-musl/x86_64-linux-musl/sys-root/usr/lib/crti.o
[20:17:54] /opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/5.2.0/crtbegin.o
[20:17:54] CMakeFiles/bin_projsync.dir/apps/projsync.cpp.o
[20:17:54] ../lib/libproj.so.21.1.2
[20:17:54] /workspace/destdir/lib/libsqlite3.so
[20:17:54] /workspace/destdir/lib/libtiff.so
[20:17:54] /workspace/destdir/lib/libcurl.so
[20:17:54] -lstdc++ (/opt/x86_64-linux-musl/x86_64-linux-musl/lib64/libstdc++.so)
[20:17:54] -lgcc_s (/opt/x86_64-linux-musl/x86_64-linux-musl/lib64/libgcc_s.so)
[20:17:54] -lc (/opt/x86_64-linux-musl/x86_64-linux-musl/sys-root/usr/lib/libc.so)
[20:17:54] -lgcc_s (/opt/x86_64-linux-musl/x86_64-linux-musl/lib64/libgcc_s.so)
[20:17:54] /opt/x86_64-linux-musl/bin/../lib/gcc/x86_64-linux-musl/5.2.0/crtend.o
[20:17:54] /opt/x86_64-linux-musl/x86_64-linux-musl/sys-root/usr/lib/crtn.o
[20:17:54] make[2]: Leaving directory '/workspace/srcdir/proj-7.2.1/build'
[20:17:54] [100%] Built target bin_projsync
```
The compiler invocation had an explicit reference to libcurl here, and all executables directly needed curl:
```console
% for file in PROJ.v700.202.100.x86_64-linux-musl-cxx11/bin/*; do echo $file; readelf -d $file | grep NEEDED | grep curl; done 
PROJ.v700.202.100.x86_64-linux-musl-cxx11/bin/cct
 0x0000000000000001 (NEEDED)             Shared library: [libcurl.so.4]
PROJ.v700.202.100.x86_64-linux-musl-cxx11/bin/cs2cs
 0x0000000000000001 (NEEDED)             Shared library: [libcurl.so.4]
PROJ.v700.202.100.x86_64-linux-musl-cxx11/bin/geod
 0x0000000000000001 (NEEDED)             Shared library: [libcurl.so.4]
PROJ.v700.202.100.x86_64-linux-musl-cxx11/bin/gie
 0x0000000000000001 (NEEDED)             Shared library: [libcurl.so.4]
PROJ.v700.202.100.x86_64-linux-musl-cxx11/bin/proj
 0x0000000000000001 (NEEDED)             Shared library: [libcurl.so.4]
PROJ.v700.202.100.x86_64-linux-musl-cxx11/bin/projinfo
 0x0000000000000001 (NEEDED)             Shared library: [libcurl.so.4]
PROJ.v700.202.100.x86_64-linux-musl-cxx11/bin/projsync
 0x0000000000000001 (NEEDED)             Shared library: [libcurl.so.4]
```
Also on `x86_64-linux-gnu` PROJ 7.2 executables used to depend on libcurl.so directly, while in 8.2 they don't (and probably don't directly need to, as libcurl is only an indirect dependency).

I still have no clue why this is a link-time error only on `x86_64-linux-musl`, but I'd say that forcing `-lcurl` on this platform only wouldn't be a big problem, it'd be basically the same as it was in 7.2

**Comment by visr** [comment]
Hmm yeah I wonder what is special about `x86_64-linux-musl` here. Isn't that the same as the host, so the only platform that is not cross compiling?

I tried the approach of https://github.com/JuliaPackaging/Yggdrasil/pull/4097#issuecomment-999883633 but it doesn't work, not sure what the right approach to force `-lcurl` is.

**Comment by giordano** [comment]
> Isn't that the same as the host

Yes, but for the most part we try not to make it special, but I don't understand why that'd make any difference here.

> so the only platform that is not cross compiling? 

No, technically we do native compilation also for i686 and x86_64 Linux with glibc.

> not sure what the right approach to force -lcurl is.

Exporting `LDFLAGS` should work.

**Pull request merged**